File: src/lib/chat/conversations.ts
export type ConversationHistoryMessage = {
	role: 'user' | 'assistant' | 'system';
	body: string;
	created_at?: string | null;
};

export type ActiveConversationMessage = ConversationHistoryMessage & {
	role: 'user' | 'assistant';
};

export const CONVERSATION_SUMMARY_PREFIX = 'Previous conversation summary:\n';

export function isConversationSummaryMessage(
	chat: Pick<ConversationHistoryMessage, 'role' | 'body'>
) {
	return chat.role === 'system' && chat.body.startsWith(CONVERSATION_SUMMARY_PREFIX);
}

export function buildConversationSummaryBody(summary: string) {
	return `${CONVERSATION_SUMMARY_PREFIX}${summary.trim()}`;
}

export function extractConversationSummary(body: string) {
	if (!body.startsWith(CONVERSATION_SUMMARY_PREFIX)) {
		return body.trim();
	}

	return body.slice(CONVERSATION_SUMMARY_PREFIX.length).trim();
}

export function partitionChatHistory(chatHistory: ConversationHistoryMessage[]) {
	const previousSummaries: string[] = [];
	let activeMessages: ActiveConversationMessage[] = [];

	for (const chat of chatHistory) {
		if (isConversationSummaryMessage(chat)) {
			previousSummaries.push(extractConversationSummary(chat.body));
			activeMessages = [];
			continue;
		}

		if (chat.role === 'user' || chat.role === 'assistant') {
			activeMessages.push({
				...chat,
				role: chat.role
			});
		}
	}

	return { previousSummaries, activeMessages };
}

export function hasActiveConversation(chatHistory: ConversationHistoryMessage[]) {
	return partitionChatHistory(chatHistory).activeMessages.length > 0;
}

export function formatConversationTranscript(chatHistory: ConversationHistoryMessage[]) {
	return chatHistory
		.filter(
			(chat): chat is ActiveConversationMessage => chat.role === 'user' || chat.role === 'assistant'
		)
		.map((chat) => `${chat.role === 'user' ? 'Human' : 'Assistant'}: ${chat.body}`)
		.join('\n');
}

File: backend/src/app/services/conversations.py
"""Port of `src/lib/chat/conversations.ts`. Pure functions operating on anything with `role`/`body`
attributes (in practice, `db.models.Chat` rows) — used by `app/chats/service.py` to separate a
persisted chat history into prior conversation summaries plus the current, live segment."""

from collections.abc import Sequence
from typing import Protocol

CONVERSATION_SUMMARY_PREFIX = "Previous conversation summary:\n"


class ConversationMessage(Protocol):
    role: str
    body: str


def is_conversation_summary_message(message: ConversationMessage) -> bool:
    return message.role == "system" and message.body.startswith(CONVERSATION_SUMMARY_PREFIX)


def build_conversation_summary_body(summary: str) -> str:
    return f"{CONVERSATION_SUMMARY_PREFIX}{summary.strip()}"


def extract_conversation_summary(body: str) -> str:
    if not body.startswith(CONVERSATION_SUMMARY_PREFIX):
        return body.strip()
    return body[len(CONVERSATION_SUMMARY_PREFIX) :].strip()


def partition_chat_history(
    chat_history: Sequence[ConversationMessage],
) -> tuple[list[str], list[ConversationMessage]]:
    """Returns `(previous_summaries, active_messages)` — `active_messages` is reset every time a
    summary marker is encountered, so it ends up holding only the most recent, still-live segment
    of the conversation."""
    previous_summaries: list[str] = []
    active_messages: list[ConversationMessage] = []

    for message in chat_history:
        if is_conversation_summary_message(message):
            previous_summaries.append(extract_conversation_summary(message.body))
            active_messages = []
            continue
        if message.role in ("user", "assistant"):
            active_messages.append(message)

    return previous_summaries, active_messages


def has_active_conversation(chat_history: Sequence[ConversationMessage]) -> bool:
    _, active_messages = partition_chat_history(chat_history)
    return len(active_messages) > 0


def format_conversation_transcript(chat_history: Sequence[ConversationMessage]) -> str:
    lines = [
        f"{'Human' if message.role == 'user' else 'Assistant'}: {message.body}"
        for message in chat_history
        if message.role in ("user", "assistant")
    ]
    return "\n".join(lines)
